### ISSUE TITLE
Support EventStore Projection

### ISSUE THREAD

**Issue by Xatter** [jet/propulsion#8]
For EventStore storage specifically, it looks like this library is designed so that EventStore is only used for storage of events and that subsequent communicating that events occured is offloaded to Kafka. The EventStore product does support messaging via PersistentSubscription and it would be nice for smaller projects to be able to use just EventStore and not have to spin up a Kafka cluster if that kind of scale is not needed.

Specifically I'd love an API where you can subscribe either to a specific aggregate `user-1234` or a projection `user-*` and have the library handle things like checkpointing (storing the last read offset for a client in EventStore itself) similar to how PersistentSubscriptions work in the EventStore API today.

**Comment by bartelink** [comment]
Looking like we'll build this - idea is to provide a dotnet-templates template which consumes from $all, and maintains offset in an Equinox.Cosmos instance. Clearly this makes no sense for apps that don't already have a dependency on such a store. With a bit of hacking, it should be possible to hook this to store it elsewhere. Interested to hear what sort of backing stores people have in mind (realistically, I personally won't be committing to actually doing the implementation)

**Comment by CumpsD** [comment]
This is an interesting issue. I'm using ES and want to project to SQL Server/Elasticsearch/.... by trailing $all.

In the past I've had this kind of functionality with SQL Stream Store and Projac to project events into SQL Server tables, but I didn't find anything similar for ES. Initially I started writing it myself, but if propulsion would support this, that would be awesome.

**Comment by bartelink** [comment]
It's looking like my work needs mean I'll be writing up a proper end to end set of docs for Propulsion that explain how its pipeline fits together soon (_I edited and deleted forward looking statements here_).

There are similar systems out there which have similar architectures for other platforms (the name escapes me) - it should be mentioned that the split of the concepts in Propulsion derives from scaling and perf tuning of large sync and projection pipelines with real data and needs (mainly mixing and matching ES and Cosmos) - i.e. this is not me trying to build some abstract projection framework.

### Sources+checkpoints

the summaryProjector template has wiring to read:
a) ES $all - with checkpoints saved in Equinox.Cosmos (writes an event every hour, updates a snapshot every 5s) _Why? because in apps I'm working on, they always have an Equinox.Cosmos store to hand anyway_
b) CosmosDB CFP - checkpoints managed by CFP (updates the checkpoint documents every 5s but there is no record)

An easy enough to achieve extension would be to make an adjusted version of (a) that writes to an EventStore stream every 5s _that has been configured to only maintain the most recent event_ (it's pretty pluggable, if you e.g. had something like Consul, you could write to that too)

_this is the main open question that is critical to resolving this Issue as I see it_

### Schedulers / Handlers

- `dotnet new summaryProjector` shows how to
    - accumulate events, maintaining order within each stream
    - handle the buffered spans per stream (up to a specified max degree of parallelism)
- `dotnet new proProjector` shows how to do various more advanced streamwise projections (only from CFP, but it works the same way)

_I need to write docs on this_

### Propulsion Sinks

Propulsion has Sinks for ES and Cosmos. Sinks take event streams and replicate them - i.e. if you are migrating from ES->ES, ES->Cosmos, Cosmos->ES or Cosmos->Cosmos - the `dotnet new proSync` template demonstrates this. One could make a `Propulsion.SqlStreamStore` that would have the same facilities (syncing from/to ES/Cosmos/SqlStreamStore into or out of SqlStreamsStore)

_this is an aside to your question, and the OP_

### ElasticSearch / SQL 

TL;DR Its conceivable that one could Projac-ish helpers in a `Propulsion.Sql` module and a similar `Propulsion.ElasticSearch` helper - but that would not be a "sink" as such - from Propulsion's point of view, its really all just schedulers running handlers, and there's no real reason for general helpers that are not reading/writing/forwarding events to live in Propulsion.

I'll expand a bit in case that helps with conveying the thinking a bit:

Internally, we don't have any work in the near term writing to Elastic Search (don't get me wrong, we use it, just not that much in projections on the team I work on atm).

On Azure, for cases where you're targeting SQL but are not doing lots of indexing and/or searching/random queries, I'd actually give sinking to `Equinox.Cosmos` a try - `dotnet new proTrackingConsumer` and `dotnet new proSummaryConsumer` show the way - if you're projecting to whats effectively a document, that's actually very efficient from a cached reading perspective.

For Sql, we do (we have internal closed source helpers which are pretty generic - i.e. they might cover Projac-like usage, but are not as purposefully designed). I dont have plans/anticipate it crossing my radar to extract/polish such a facility, but I do know that lots of folks would use it. If you had a free standing set of helpers, a `Propulsion.Sql` module might not be crazy though - the first thing to make it do would be to port `dotnet new proSummaryConsumer` to it. Not sure if it would make that much sense to actually tie its releases and/or development to this repo though - ultimaltely Propulsion majors on:
- sources from Kafka/Cosmos CFP/SqlStreamStore/EventStore
- with parallel/streamwise/multi-stream scheduling strategies
- invoking handlers that do things
  - special case: if that thing is "syncing events", then the logic to do that can sometimes make sense to host in Propulsion, e.g. Propulsion.EventStore has logic for syncing events to ES, Propulsion.Cosmos has some helpers for syncing events, but in general delegates to Equinox.Cosmos for the actual writing)

**Comment by bartelink** [comment]
related: [this (incomplete) PR](https://github.com/jet/dotnet-templates/pull/81) provides for projecting from ESDB $all, storing checkpoints in ESDB

**Comment by bartelink** [comment]
related: https://github.com/absolutejam/Propulsion.EventStoreDB

**Comment by bartelink** [comment]
related: Propulsion.EventStoreDb in V4 implements the reader using gRPC. Leaving this issue open as there's no in-the box checkpointing that stores the checkpoints in ESDB (supported checkpoint stores are DynamoDB, CosmosDB and SQL Server)
